crates/fiber-lib/src/store/pub_sub/subscription.rs

```diff
@@ -0,0 +1,144 @@
+//! Publish-subscribe notifications for the store.
+
+use std::sync::Arc;
+
+use ractor::{port::OutputPortSubscriber, OutputPort};
+
+use crate::{
+    fiber::{payment::PaymentStatus, types::Hash256},
+    invoice::CkbInvoiceStatus,
+};
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum InvoiceUpdatedPayload {
+    /// The invoice is open and can be paid.
+    Open,
+    /// The invoice is cancelled.
+    Cancelled,
+    /// The invoice is expired.
+    Expired,
+    /// The invoice is received, but not settled yet.
+    Received {
+        /// The amount of the invoice.
+        amount: u128,
+        /// Depending on whether AMP is supported, the invoice may have multiple parts,
+        /// this field indicates if we received all parts.
+        is_finished: bool,
+    },
+    /// The invoice is paid.
+    Paid,
+}
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct InvoiceUpdatedEvent {
+    pub invoice_hash: Hash256,
+    pub payload: InvoiceUpdatedPayload,
+}
+
+impl TryFrom<CkbInvoiceStatus> for InvoiceUpdatedPayload {
+    type Error = String;
+
+    fn try_from(value: CkbInvoiceStatus) -> Result<Self, Self::Error> {
+        let payload = match value {
+            CkbInvoiceStatus::Open => InvoiceUpdatedPayload::Open,
+            CkbInvoiceStatus::Cancelled => InvoiceUpdatedPayload::Cancelled,
+            CkbInvoiceStatus::Expired => InvoiceUpdatedPayload::Expired,
+            CkbInvoiceStatus::Received => {
+                return Err("InvoiceUpdatedPayload::Received requires extra data".to_string())
+            }
+            CkbInvoiceStatus::Paid => InvoiceUpdatedPayload::Paid,
+        };
+        Ok(payload)
+    }
+}
+
+impl InvoiceUpdatedEvent {
+    pub fn new(invoice_hash: Hash256, payload: InvoiceUpdatedPayload) -> Self {
+        Self {
+            invoice_hash,
+            payload,
+        }
+    }
+}
+
+// but with additional information for downstream services.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum PaymentUpdatedPayload {
+    /// initial status, payment session is created, no HTLC is sent
+    Created,
+    /// the first hop AddTlc is sent successfully and waiting for the response
+    Inflight,
+    /// related HTLC is successfully settled
+    Success { preimage: Hash256 },
```

Comment:
for MPP/AMP, how to specify this preimage field?
now after merging MPP, one payment may have multiple payment attempts, for basic MPP they share the same preimage,
but for AMP, different payment attempts using different preimages.

Reply:
CCC supports only single path now. MPP/AMP needs extra works since lnd should also use MPP/AMP and the same number of paths.